plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/ViewportPluginImpl.kt

```diff
@@ -0,0 +1,331 @@
+   *
+   * @param targetState The target [ViewportState] to transition to.
+   * @param completionListener The listener to observe the completion state.
+   */
+  override fun transitionTo(targetState: ViewportState, completionListener: CompletionListener?) {
+    with(status) {
+      when (this) {
+        is ViewportStatus.State -> {
+          if (this.state === targetState) {
+            completionListener?.onComplete(true)
+            return
+          }
+        }
+        is ViewportStatus.Transition -> {
+          if (this.toState === targetState) {
+            completionListener?.onComplete(false)
+            return
+          }
+        }
+      }
+    }
+    val fromState = status.getCurrentOrTargetState()
+    currentCancelable?.cancel()
+    currentCancelable = null
+
+    // get the transition (or default) for the from and to state
+    val transition = getTransition(fromState, targetState) ?: defaultTransition
+
+    // run the transition
+    var completionBlockInvoked = false
+    val transitionCancelable = transition.run(fromState, targetState) { isFinished ->
+      completionBlockInvoked = true
+      if (isFinished) {
+        // transfer camera updating responsibility to targetState
+        currentCancelable = targetState.startUpdatingCamera()
+        updateStatus(
+          ViewportStatus.State(targetState),
+          VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+        )
+      }
+      completionListener?.onComplete(isFinished)
+    }
+    // since it's possible that a transition might invoke its
+    // completion block synchronously, we'll only store the
+    // transition cancelable if the transition is not complete
+    // so that we don't clobber the targetState cancelable.
+    if (!completionBlockInvoked) {
+      currentCancelable = transitionCancelable
+      updateStatus(
+        ViewportStatus.Transition(transition, fromState, targetState),
+        VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC
+      )
+    }
+  }
+
+  /**
+   * Immediately goes to IDLE state canceling all ongoing transitions.
+   */
+  override fun idle() {
+    if (status.getCurrentOrTargetState() == null) return
+    currentCancelable?.cancel()
+    currentCancelable = null
+    updateStatus(ViewportStatus.State(null), VIEWPORT_STATUS_OBSERVER_REASON_PROGRAMMATIC)
+  }
+
+  /**
+   * Options that impact the [ViewportPlugin].
+   */
+  override var options: ViewportOptions = ViewportOptions.Builder().build()
+
+  private fun notifyStatusChanged(
+    previousStatus: ViewportStatus,
+    currentStatus: ViewportStatus,
+    reason: String
+  ) {
+    registeredStatusObservers.forEach {
+      it.onViewportStatusChanged(
+        previousStatus,
+        currentStatus,
+        reason
+      )
+    }
+  }
+
+  /**
+   * DefaultViewportTransition with default options
+   *
+   * This transition is used unless overridden by one of the registered transitions.
+   */
+  override lateinit var defaultTransition: ViewportTransition
```

Comment:
couldn't it lead to an easy reproducible crash if this variable gets accessed if it was not initialized?

Reply:
There's no easy way around it, as the initialisation of DefaultViewportTransition need to be async with delegate provider. Also as these plugins are initialised and onDelegateProvider are available already inside the MapView's constructor, when users access plugin from `mapView.viewport`, the defaultTransition should be ready to used.

Reply:
In v11 I think we could revisit the plugin setup and inject the required delegateProvider etc in the plugin implementation's constructor, as we got rid of the reflections and don't require an empty constructor.